### PR TITLE
fix(#7629): validate map-valued properties consistently across CREATE, MERGE, and SET

### LINKED ISSUE #7629: openCypher SET rejects map-valued properties (regression; CREATE and SQL UPDATE still accept them)
**Title:** openCypher `SET` rejects map-valued properties (regression; `CREATE` and SQL `UPDATE` still accept them)

### ArcadeDB Version:

`26.9.1` (regression; `26.7.1` and `26.8.1` are both unaffected)

### OS and JDK Version:

`Linux 7.0.12-linuxkit - OpenJDK 64-Bit Server VM 21.0.12 (Temurin-21.0.12+8)`

### Expected behavior

`SET n += $p` (and `SET n = $p`) should accept a property value that is a map, the same way `CREATE` and SQL `UPDATE` already do.

### Actual behavior

From `26.9.1` onward, the openCypher `SET` clause refuses any property whose value is a map, or a list containing maps. `CREATE` still accepts both, and SQL `UPDATE` still accepts both, so the same value is writable on insert but not on update — there is no openCypher spelling that updates a map property on an existing record.

Client-side (Bolt), the error is only:
```
code: Neo.DatabaseError.General.UnknownError
message: "Error executing Cypher command: MATCH ... SET n += $p"
```
The real cause is only in the server log:
```
com.arcadedb.exception.CommandExecutionException: Error executing Cypher command: ...
Caused by: java.lang.IllegalArgumentException: TypeError: InvalidPropertyType - Property values can not be maps
    at com.arcadedb.query.opencypher.executor.steps.SetClauseApplier.validatePropertyValue(SetClauseApplier.java:556)
    at com.arcadedb.query.opencypher.executor.steps.SetClauseApplier.coerceAndValidate(SetClauseApplier.java:543)
    at com.arcadedb.query.opencypher.executor.steps.SetClauseApplier.toPropertyMap(SetClauseApplier.java:407)
    at com.arcadedb.query.opencypher.executor.steps.SetClauseApplier.apply(SetClauseApplier.java:177)
    at com.arcadedb.query.opencypher.executor.steps.SetStep.applySetOperations(SetStep.java:137)
```
A list containing maps reports the sibling message, `Property values can not contain map values`, from the same class.

Bisected against consecutive releases: `26.7.1` and `26.8.1` both work; `26.9.1` fails. Traced to #6863, which unifies `SET`/`MERGE` into one `SetClauseApplier` and — per its own description — extends the existing scalar-property validation to every entry of a map right-hand side, closing what the PR calls "the way around the check". That closes a real gap for `SET n.x = {y: 1}`-style nested maps, but as a side effect it also closes `SET n = $map` / `SET n += $map` for a map whose value happens to be another map — which `CREATE` and SQL `UPDATE` still allow at the top level. Given `CREATE`/`UPDATE` treat it as a normal value, this looks like an unintended side effect of that validation extending further than the entity-vs-map right-hand-side check the PR describes, rather than an intended tightening.

### Steps to reproduce

No schema required — the property is undeclared in this reproduction.

```javascript
import neo4j from 'neo4j-driver'

const d = neo4j.driver('bolt://localhost:7687', neo4j.auth.basic('root', '<password>'))
const s = d.session({ database: 'setrepro' })   // an empty database is enough

const MAP = { k: 1, nested: { deep: true } }

await s.run('CREATE (n:R {id: 1, m: $m})', { m: MAP })            // ok on 26.7.1 and 26.9.1
await s.run('MATCH (n:R {id: 1}) SET n += $p', { p: { s: 'ok' } }) // ok on both — scalars

await s.run('MATCH (n:R {id: 1}) SET n += $p', { p: { m: MAP } }) // ok on 26.7.1/26.8.1, FAILS on 26.9.1
```

Given `CREATE` and SQL `UPDATE` both accept the same value, the storage layer clearly supports it. #6863's own tests (`CypherMergeSetClauseParityIssue6831Test`, `CypherSetFromEntityIssue6832Test`) don't cover a map-valued map entry, so this looks like an untested side effect of that PR rather than something it deliberately decided.

### PR DESCRIPTION
**Pull request by lvca**
## Summary

Closes #7629, though not the way it was originally framed - see the analysis in the issue thread. Short version:

- **SET's rejection was already correct.** Real Neo4j refuses any map-valued property ("Property values can only be of primitive types or arrays thereof"), and `SetClauseApplier` already enforced this since #6863, with 5 tests covering the literal-map shapes. The reported repro (`SET n += $p` with `p.m` a map) is the same rule, just reached through a bound parameter instead of a literal.
- **The actual bug:** `CreateStep` and `MergeStep`'s creation branch never applied this check at all, so `CREATE (n {m: $m})` with a map `$m` silently stored it - the real CREATE/SET inconsistency.
- **The fix:** extracted the check into `CypherValues.coerceAndValidatePropertyValue` (shared with `SetClauseApplier`, removing its private duplicate) and applied it to every property-write path in `CreateStep` and `MergeStep`'s creation branch - vertex and edge properties, map-literal and bare-parameter forms.
- **A second bug found and fixed during review:** `point()` represents a Point as a plain `Map` internally (ArcadeDB has no dedicated Geometry runtime type yet, #4870), so the new check also rejected Point-valued properties - which Neo4j treats as a primitive, not a map. `CypherValues.isPointShaped` recognizes one structurally (numeric `x`/`y` and a non-null `crs`, the same keys every branch of `CypherPointFunction` already writes and other geo functions already key off), so it survives a storage round-trip (unlike an identity-based check, which broke as soon as a Point property was copied from an existing node). A point-shaped map is exempt as a whole, but every one of its own entries is still validated, so a nested map/list can't be smuggled in under some other key.
- **A third gap found during review:** `MergeStep.createEdge` didn't skip a `null`-valued property the way the vertex creation branch (`setProperties`) does, so `MERGE (a)-[r:REL {p: $x}]->(b)` with `$x` null stored an explicit null on the edge while the vertex equivalent didn't. Normalized.
- **A desirable side effect worth calling out:** because `SetClauseApplier` now delegates to the same shared `CypherValues.coerceAndValidatePropertyValue`, SET also picks up the `isPointShaped` exemption. Before this PR, `SET n.loc = point(...)` threw `TypeError: InvalidPropertyType` on `main` (the old private check rejected any `Map` unconditionally) - the mirror image of the CREATE/MERGE gap above. SET now accepts a Point-valued property exactly like CREATE and MERGE do.
- **A pre-existing test fixture bug this PR's stricter validation exposed:** the shared Bolt conformance fixture (`bolt/conformance/fixtures/type-matrix.cypher`, issue #4883, used by all 5 Bolt e2e client suites) seeded a node with a literal nested map property as part of TYPE-005 "nested lists and maps round-trip structurally". A map-valued property was never actually valid Cypher (real Neo4j refuses it too), so the fixture was asserting behavior that shouldn't have been claimed as conformant. Dropped the map property from the fixture and renamed the scenario to "Nested lists round-trip structurally" (`spec.yaml` + the java/js/python/csharp/go test suites).

SQL `UPDATE` accepting map values is unrelated and correct as-is - SQL is ArcadeDB's own multi-model query language with a first-class `MAP` schema type, not bound by Neo4j/TCK type rules.

## Test plan

- `CypherMapPropertyConsistencyIssue7629Test` (17 tests): CREATE/MERGE reject map-valued vertex and edge properties (map-literal and bare-parameter forms), scalars/lists still work, a `point()`-valued property still stores and round-trips through storage correctly on CREATE, SET now also accepts a Point-valued property (dot-property and merge-map forms), a point-shaped map can't smuggle a nested map under another key, and a null-valued property is skipped consistently on both the vertex and edge MERGE creation branches.
- Full openCypher TCK suite: 0 failures (unchanged from baseline; `clauses/create`, `clauses/merge`, `clauses/set` all 100%).
- Full `com.arcadedb.query.opencypher` test package: 9028 tests, 0 failures/errors.
- Existing `CypherMergeSetClauseParityIssue6831Test` / `CypherSetFromEntityIssue6832Test` (the #6863 regression suite): unchanged, still green.
- The spatial function suites (`OpenCypherSpatialFunctionsTest`, `OpenCypherSpatialFunctionsComprehensiveTest`, `FunctionCachingTest`) that exercise `point()` storage: unchanged, still green.

## Files changed
- `engine/src/main/java/com/arcadedb/query/opencypher/executor/CypherValues.java` - shared `coerceAndValidatePropertyValue`/`isPointShaped` validation
- `engine/src/main/java/com/arcadedb/query/opencypher/executor/steps/SetClauseApplier.java` - delegates to the shared check instead of its own private copy
- `engine/src/main/java/com/arcadedb/query/opencypher/executor/steps/CreateStep.java` - all property-write paths (vertex/edge, map-literal/bare-parameter) now validate
- `engine/src/main/java/com/arcadedb/query/opencypher/executor/steps/MergeStep.java` - creation branch (vertex/edge) now validates and consistently skips null values
- `engine/src/test/java/com/arcadedb/query/opencypher/CypherMapPropertyConsistencyIssue7629Test.java` (new)
- `bolt/conformance/fixtures/type-matrix.cypher`, `bolt/conformance/spec.yaml` - dropped the invalid nested-map fixture property, renamed TYPE-005
- `e2e/src/test/java/com/arcadedb/e2e/RemoteBoltDatabaseIT.java`, `e2e-js/src/js-bolt-conformance.test.js`, `e2e-python/tests/test_bolt.py`, `e2e-csharp/ArcadeDB.E2ETests/BoltE2ETests.cs`, `e2e-go/bolt_conformance_test.go` - matched the fixture change

## Summary by CodeRabbit

* **Bug Fixes**
  * Property values are now consistently validated across CREATE, MERGE, and SET operations.
  * Unsupported map-valued properties, including nested maps, are rejected consistently.
  * Null-valued properties are omitted during MERGE creation.
  * Valid point values and nested lists continue to be supported.

* **Conformance**
  * Type round-trip checks now cover nested lists without claiming support for map-valued properties.

### CODE REVIEW

File: engine/src/test/java/com/arcadedb/query/opencypher/CypherMapPropertyConsistencyIssue7629Test.java [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+    assertThatThrownBy(() -> database.transaction(() -> database.command("opencypher",
+        "MATCH (a:R {id: 1}), (b:R {id: 2}) CREATE (a)-[r:REL {m: $m}]->(b)", Map.of("m", Map.of("k", 1)))))
+        .rootCause()
+        .hasMessageContaining("TypeError: InvalidPropertyType");
+  }
+
+  @Test
+  void createStillAcceptsOrdinaryScalarAndListParameters() {
+    database.transaction(() -> database.command("opencypher",
+        "CREATE (n:R {id: 1, s: $s, tags: $tags})", Map.of("s", "ok", "tags", List.of("x", "y"))));
```
[the remark this answers is not in the document]
Comment (lvca) [reply engine/src/test/java/com/arcadedb/query/opencypher/CypherMapPropertyConsistencyIssue7629Test.java:134]: Added: `createRejectsAListContainingAMapParameter` now covers CREATE with `tags: [{k: 1}]`-shaped list-of-maps and asserts `TypeError: InvalidPropertyType`. Pushed in 27cc7b14d.

**Comment by lvca** [comment]
Addressed all three points from the review (27cc7b14d):

1. **The blocking bug (Point identity vs. round-trip)**: confirmed and fixed. Switched from checking `instanceof CypherPoint` to a structural check - a map carrying the `x`/`y`/`crs` keys every branch of `CypherPointFunction` writes (the same keys `CypherPointDistanceFunction`/`PointWithinBBoxFunction` already key off to read one back). Removed the now-unnecessary `CypherPoint` class. Added a regression test for exactly the round-trip scenario called out (`MATCH (a) CREATE (b {loc: a.loc})` across separate transactions).
2. **`MergeStep.setProperties` null handling**: normalized to skip `null` before validating, matching `CreateStep`. No functional change (both paths already ended up storing `null` either way), just consistency while the code was already touched.
3. **List-of-Points / list-of-maps test coverage**: added both - `createAcceptsAListOfPoints` and `createRejectsAListContainingAMapParameter` (the latter also requested by CodeRabbit).

Not changing the `Object[]` vs `List` note in `validatePropertyValue` - agreed it's moot today (Cypher list values are always `java.util.List` in this codebase) and speculative hardening for a shape that can't currently occur isn't worth the code per this repo's guidelines; happy to revisit if that ever changes.

**Comment by lvca** [comment]
Thanks - addressed in e98c70804:

1. **Point-shaped-map loophole**: agreed, valid. Tightened \`isPointShaped\` to require \`x\`/\`y\` are actually \`Number\`s, and more importantly, changed the exemption so it only waives the "is a Map" check on the point-shaped map itself - every one of its own entries is still run through \`validatePropertyValue\`. So \`{x: 1, y: 2, crs: 'x', payload: {secret: 1}}\` is now refused (the \`payload\` map is caught), while a real \`point()\`-produced map (whose values are only Numbers/Strings) passes through unaffected. Added \`createRejectsAPointShapedMapSmugglingANestedMap\` to pin it.
2. **Test name typo**: fixed (\`Earlierly\` → \`Previously\`).
3. **\`MergeStep.evaluateProperties\` redundant \`toCoreJavaType\` call**: leaving this as-is - you already flagged it as pre-existing (not introduced or worsened in any correctness-relevant way by this PR) and harmless (idempotent conversion). Fixing it would mean touching \`evaluateProperties\`'s per-branch conversion logic, which is a separate cleanup from "make property-value validation consistent across CREATE/MERGE/SET" and outside this PR's scope.

Verified against the full \`com.arcadedb.query.opencypher\` test package again (9023 tests, 0 failures) plus the spatial suites.

File: engine/src/main/java/com/arcadedb/query/opencypher/executor/CypherValues.java [hunk 2]
```diff
@@ -43,4 +49,46 @@ public static boolean equalValues(final Object a, final Object b) {
...
+          throw new IllegalArgumentException("TypeError: InvalidPropertyType - Property values can not contain map values");
+        if (element instanceof List)
+          validatePropertyValue(element);
+      }
+    } else if (value instanceof Map map && !isPointShaped(map))
+      throw new IllegalArgumentException("TypeError: InvalidPropertyType - Property values can not be maps");
+  }
+
+  private static boolean isPointShaped(final Map<?, ?> map) {
+    return map.containsKey("crs") && map.containsKey("x") && map.containsKey("y");
```
[the remark this answers is not in the document]
Comment (lvca) [reply engine/src/main/java/com/arcadedb/query/opencypher/executor/CypherValues.java]: Already fixed in e98c70804 (pushed before this review posted): isPointShaped now requires x/y to be Number, and the exemption only waives the map-check on the point-shaped map itself - every entry is still recursively validated, so a nested `payload: {secret: 1}` is caught (pinned by `createRejectsAPointShapedMapSmugglingANestedMap`).

Went with recursive value-validation rather than a strict key whitelist: it closes the same hole (no map/list-of-maps can hide anywhere in the structure) without hardcoding CypherPointFunction's exact key set, so it stays correct if that function ever adds a field. Also skipping the 'crs must be a string' requirement - your own note points out the Cartesian branch has no CRS-value contract, and a non-Map crs value isn't a data-integrity risk the way a smuggled map is.

**Comment by lvca** [comment]
Addressed in ee2b3a5b9:

1. **\`MergeStep.createEdge\` null-skip gap**: confirmed and fixed - added the same \`if (value != null)\` guard used in \`setProperties\`. Added regression tests for both the vertex (\`mergeCreationBranchSkipsANullVertexProperty\`) and edge (\`mergeCreationBranchSkipsANullEdgeProperty\`) creation-branch cases.
2. **\`isPointShaped\` heuristic tradeoff**: agreed this is a known, documented limitation rather than something to fix now - no dedicated Geometry type exists yet (#4870), and you already noted it's not blocking. Left as-is; the javadoc on \`coerceAndValidatePropertyValue\` covers the tradeoff.
3. **\`csharp-e2e-tests\`/\`go-e2e-tests\` CI failures**: investigated - both fail during environment setup with a Docker image \`digest-mismatch\` and generic HTTP 400s across every single test in the suite (auth, protocol negotiation, unrelated type round-trips), before any query specific to this PR would even run. Checked recent \`main\` branch runs of the same workflow: they're currently red too, just in different jobs (\`ha-integration-tests\`, \`unit-tests\`, \`integration-tests\`) - consistent with this being pre-existing CI infrastructure flakiness, not something this PR's diff caused.

Verified against the full \`com.arcadedb.query.opencypher\` test package again (9025 tests, 0 failures).

**Comment by lvca** [comment]
Addressed in 64ab7264f, plus the description:

1. **Stale PR description**: updated - removed the mention of \`CypherPointFunction.java\`/\`CypherPoint.java\`, which the second commit's switch to structural detection made unnecessary; the description now lists the actual final file set and mentions all three review-driven fixes (Point round-trip, smuggling loophole, null-skip parity).
2. **\`isPointShaped\` heuristic comment**: added a method-level javadoc explaining the tradeoff explicitly (known, narrow escape hatch pending #4870's dedicated Geometry type - not fixable today, not a bug in this PR).
3. **Bare-parameter CREATE form untested**: added \`createRejectsAMapValuedPropertyViaBareParameterSyntax\` and its edge equivalent, confirming \`setPropertiesFromParameter\`/\`buildPropertiesFromParameter\` funnel through the same validation as the map-literal form.
4. **\`crs\` non-null nit**: fixed - \`isPointShaped\` now requires \`map.get("crs") != null\`, not just \`containsKey\`.

Verified against the full \`com.arcadedb.query.opencypher\` test package again (9027 tests, 0 failures).

@lvca this should be ready for a final look/merge - four review rounds (claude-review + CodeRabbit) have all been addressed or explicitly judged not worth doing (CodeRabbit itself withdrew its point-shaped-map finding after re-verifying the fix), and the two e2e CI failures are confirmed pre-existing infra flakiness unrelated to this change.

**Comment by lvca** [comment]
Great catch - confirmed and addressed in 7b2da8f6a:

Added \`setAcceptsAPointValuedPropertyLikeCreateAndMerge\`, pinning both \`SET n.loc = point(...)\` (dot-property form) and \`SET n += {loc: point(...)}\` (merge-map form) now succeed, where they'd have thrown \`TypeError: InvalidPropertyType\` on \`main\` before this PR (the old private \`validatePropertyValue\` in \`SetClauseApplier\` rejected any \`Map\` unconditionally). No production code change needed - this was already correct as of the second commit (once \`SetClauseApplier\` started delegating to the shared \`CypherValues.coerceAndValidatePropertyValue\`), just untested and undocumented.

Also updated the PR description to call this out explicitly as a fifth bullet, rather than leaving it as an implicit side effect.

Verified against the full \`com.arcadedb.query.opencypher\` test package again (9028 tests, 0 failures).

**Comment by lvca** [comment]
**Correction to my earlier comment** claiming \`csharp-e2e-tests\`/\`go-e2e-tests\` were pre-existing infra flakiness - that was wrong, and I want to flag it rather than let it stand uncorrected.

The real cause (confirmed in aee599a3b): all four Bolt e2e client suites (java, js, python, csharp, go) share \`bolt/conformance/fixtures/type-matrix.cypher\` (issue #4883), which seeded a \`TypeMatrix\` node with a literal nested map property (\`nestedMapProp: {a: 1, b: {c: 2}}\`) as part of TYPE-005 ("nested lists and maps round-trip structurally"). This PR's stricter CREATE validation correctly rejects that value - a map isn't a primitive or an array of primitives, so real Neo4j would refuse it too - which broke setup for every test in all four suites, not just two. The Docker "digest-mismatch" log line I pointed to earlier is an unrelated, benign warning that appears regardless of outcome; I mistook it for the cause.

Fix: dropped \`nestedMapProp\` from the fixture (kept \`pointProp\`/\`nestedListProp\`, still valid), renamed TYPE-005 to "Nested lists round-trip structurally", and updated \`spec.yaml\` plus all five test suites (the Java e2e IT keeps its own mirrored copy of the fixture) to match. Left the generated \`COMPATIBILITY.md\` alone since it's nightly-rendered from \`spec.yaml\` and a live run, not hand-edited.

This is exactly the kind of behavior change worth surfacing explicitly rather than silently patching around: the fixture was asserting something that was never actually valid Cypher, and this PR's fix correctly caught it.

**Comment by lvca** [comment]
\`unit-tests\` failed, but with exactly 2 failures, both in \`com.arcadedb.engine.timeseries\` - nowhere near this PR's diff (openCypher CREATE/MERGE/SET property validation):

- \`Issue7089NaNTransparentSumAvgTest.oneNaNSampleNoLongerPoisonsTheBucketOnAnyPath\` - corresponds to issue #7089, which was already tracked as open/unresolved from unrelated prior work (JSONArray.put(Number) rewriting NaN to 0, #7032/#7039/#7043 batch).
- \`MultiColumnAggregationResultTest.emptySumAndCountStayZeroNotNaN\` - already documented as pre-existing red on \`main\` from an unrelated investigation (issue #7615's async-batch work, itself a side effect of the #4597 fix, unfiled).

Neither test touches Cypher, property validation, \`CypherValues\`, \`CreateStep\`, \`MergeStep\`, or \`SetClauseApplier\`. \`integration-tests\` (which runs a large, overlapping chunk of the suite) passed. Given both failures are pre-existing, already-tracked, and in a completely unrelated engine area, I'm treating this as confirmed pre-existing flakiness rather than something this PR introduced or needs to fix.

**Comment by lvca** [comment]
\`ha-integration-tests\` failed with 3 test classes, all in \`com.arcadedb.server.ha.raft\` - unrelated to this PR's diff:

- \`RaftUserManagement3NodesIT.restUserEndpointsReplicateAcrossCluster\`: \`org.awaitility.core.ConditionTimeoutException\` - a 15s condition wait timeout on cross-cluster REST replication.
- \`RaftPriorityRejoinIT.leaderRestartThenReplicaRestartConverges\`: \`ReplicationDispatchedTimeoutException\` - a Raft group-commit dispatch timeout.
- \`Issue7511MixedVersionSecurityEntryIT.aGroupOrTokenChangeIsRefusedWhileOneNodeCannotDecodeTheEntryAndResumesAfterTheUpgrade\`: an eventual-convergence assertion that didn't converge in time.

All three are timing/convergence failures in multi-node Raft cluster tests - the textbook shape of the chronic \`ha-integration-tests\` flake this CI lane is known for, not something this PR's Cypher property-validation change could cause. \`integration-tests\` (which also exercises HA to a lesser degree) passed.

## Summary of this PR's review cycle

- **Issue diagnosis**: SET's rejection was already correct/Neo4j-faithful; the real bug was CREATE/MERGE never applying the same map-property check.
- **7 rounds of \`claude-review\`** (all now clean) + **CodeRabbit** (withdrew its one finding after re-verifying the fix) found and fixed, in order: the Point-identity-vs-storage-round-trip bug, the point-shaped-map smuggling loophole, the \`MergeStep.createEdge\` null-skip asymmetry, a stale PR description, missing bare-parameter test coverage, and a \`crs\` null-check tightening.
- **A real regression found mid-review**: the stricter CREATE validation broke a shared Bolt-conformance fixture (\`type-matrix.cypher\`) across all 5 e2e client suites (java/js/python/csharp/go) - fixed by removing the invalid nested-map fixture property.
- **CI**: full matrix green except two verified-unrelated failures - \`unit-tests\` (2 pre-existing timeseries tests, one tied to already-open issue #7089) and \`ha-integration-tests\` (3 Raft timing flakes, documented above).

Ready for a final look/merge.
